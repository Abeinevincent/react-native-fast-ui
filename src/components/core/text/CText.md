# CText Component

The `CText` component is a custom text component that extends the functionality of the React Native `Text` component.

## Props

### Children Props

- `children`: (Required) Content to render on the ui.

### Style-related Props

- `w`: (Optional) Width of the text. Accepts a string or a number. Default: '100%'.
- `h`: (Optional) Height of the text. Accepts a string or a number. Default: 'auto'.
- `c`: (Optional) Text color. Accepts a string representing a color code. Default: 'black'.
- `ta`: (Optional) Text alignment. Accepts one of: 'auto', 'left', 'right', 'center', 'justify'. Default: 'auto'.
- `tav`: (Optional) Vertical text alignment. Accepts one of: 'auto', 'top', 'bottom', 'center'. Default: 'auto'.
- `ff`: (Optional) Font family. Accepts a string representing the font family. Default: 'System'.
- `fs`: (Optional) Font size. Accepts a number representing font size. Default: 14.
- `fst`: (Optional) Font style. Accepts one of: 'normal', 'italic'. Default: 'normal'.
- `fw`: (Optional) Font weight. Accepts one of: 'normal', 'bold', or a numeric value from '100' to '900'. Default: 'normal'.
- `ls`: (Optional) Letter spacing. Accepts a number representing letter spacing. Default: 0.
- `lh`: (Optional) Line height. Accepts a number representing line height. Default: 20.
- `tdl`: (Optional) Text decoration line. Accepts one of: 'none', 'underline', 'line-through', 'underline line-through'. Default: 'none'.
- `tds`: (Optional) Text decoration style. Accepts one of: 'solid', 'double', 'dotted', 'dashed'. Default: 'solid'.
- `tdc`: (Optional) Text decoration color. Accepts a string representing a color code. Default: 'black'.
- `tsc`: (Optional) Text shadow color. Accepts a string representing a color code. Default: 'transparent'.
- `tso`: (Optional) Text shadow offset. Accepts an object with `width` and `height` properties. Default: `{ width: 0, height: 0 }`.
- `tsr`: (Optional) Text shadow radius. Accepts a number representing the text shadow radius. Default: 0.
- `tt`: (Optional) Text transform. Accepts one of: 'none', 'uppercase', 'lowercase', 'capitalize'. Default: 'none'.
- `bber`: (Optional) Border bottom end radius. Accepts a number.
- `bblr`: (Optional) Border bottom left radius. Accepts a number.
- `bbrr`: (Optional) Border bottom right radius. Accepts a number.
- `bbsr`: (Optional) Border bottom start radius. Accepts a number.
- `bser`: (Optional) Border start end radius. Accepts a number.
- `bssr`: (Optional) Border start start radius. Accepts a number.
- `beer`: (Optional) Border end end radius. Accepts a number.
- `besr`: (Optional) Border end start radius. Accepts a number.
- `bbw`: (Optional) Border bottom width. Accepts a number.
- `bc`: (Optional) Border color. Accepts a string representing a color code.
- `bci`: (Optional) Border curve iOS. Accepts one of: 'circular', 'continuous'.
- `bec`: (Optional) Border end color. Accepts a string representing a color code.
- `blc`: (Optional) Border left color. Accepts a string representing a color code.
- `blw`: (Optional) Border left width. Accepts a number.
- `br`: (Optional) Border radius. Accepts a number.
- `brc`: (Optional) Border right color. Accepts a string representing a color code.
- `brw`: (Optional) Border right width. Accepts a number.
- `bsc`: (Optional) Border start color. Accepts a string representing a color code.
- `bs`: (Optional) Border style. Accepts one of: 'solid', 'dotted', 'dashed'.
- `btc`: (Optional) Border top color. Accepts a string representing a color code.
- `bter`: (Optional) Border top end radius. Accepts a number.
- `btlr`: (Optional) Border top left radius. Accepts a number.
- `btrr`: (Optional) Border top right radius. Accepts a number.
- `btsr`: (Optional) Border top start radius. Accepts a number.
- `btw`: (Optional) Border top width. Accepts a number.
- `bw`: (Optional) Border width. Accepts a number.
- `el`: (Optional) Elevation Android. Sets the elevation of the view. Accepts a number.
- `o`: (Optional) Opacity. Accepts a number.

### Margin and Padding Props

- `m`: (Optional) Margin. Accepts a number or a string representing margin value. Default: 0.
- `mt`: (Optional) Margin top. Accepts a number or a string representing margin top value. Default: 0.
- `mr`: (Optional) Margin right. Accepts a number or a string representing margin right value. Default: 0.
- `mb`: (Optional) Margin bottom. Accepts a number or a string representing margin bottom value. Default: 0.
- `ml`: (Optional) Margin left. Accepts a number or a string representing margin left value. Default: 0.
- `p`: (Optional) Padding. Accepts a number or a string representing padding value. Default: 0.
- `pt`: (Optional) Padding top. Accepts a number or a string representing padding top value. Default: 0.
- `pr`: (Optional) Padding right. Accepts a number or a string representing padding right value. Default: 0.
- `pb`: (Optional) Padding bottom. Accepts a number or a string representing padding bottom value. Default: 0.
- `pl`: (Optional) Padding left. Accepts a number or a string representing padding left value. Default: 0.

### Platform-specific Props

- `ifp`: (Optional) Include font padding (Android). Accepts a boolean. Default: true.
- `va`: (Optional) Vertical align (Android). Accepts one of: 'auto', 'top', 'bottom', 'middle'. Default: 'auto'.
- `wd`: (Optional) Writing direction (iOS). Accepts one of: 'auto', 'ltr', 'rtl'. Default: 'auto'.

### Universal Props

- `us`: (Optional) User select. Accepts one of: 'auto', 'text', 'none', 'contain', 'all'. Default: 'none'.

## Example Usage

```jsx
import { CText } from './CText';

// Example usage of CText component
<CText
  w={100}
  h={50}
  c="red"
  fontSize={20}
  fontWeight="bold"
  margin={10}
  padding={5}
>
  Custom Text
</CText>;
```
