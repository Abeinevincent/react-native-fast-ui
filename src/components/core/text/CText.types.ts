import {
  type ColorValue,
  type DimensionValue,
  type TextProps,
} from 'react-native';
import { type IReusableStyleTypes } from '../../../types/reusabletypes/ReusableStyleTypes';
import type { ReactNode } from 'react';

// Define interface for CText props extending TextProps
export interface ICText extends TextProps, IReusableStyleTypes {
  // Children
  children: ReactNode;
  // Style-related props
  c?: string; // Text color prop
  bgc?: ColorValue | string; // backgroundColor
  h?: DimensionValue | undefined;
  w?: DimensionValue | undefined;
  ta?: 'auto' | 'left' | 'right' | 'center' | 'justify'; // text align
  tav?: 'auto' | 'top' | 'bottom' | 'center'; //text align vertical
  ff?: string; // Font family prop
  fs?: number; // Font size prop
  fst?: 'normal' | 'italic'; // Font style prop
  fw?:
    | 'normal'
    | 'bold'
    | '100'
    | '200'
    | '300'
    | '400'
    | '500'
    | '600'
    | '700'
    | '800'
    | '900'; // Font weight prop
  ls?: number; // Letter spacing prop
  lh?: number; // Line height prop
  tdl?: 'none' | 'underline' | 'line-through' | 'underline line-through'; // Text decoration line prop
  tds?: 'solid' | 'double' | 'dotted' | 'dashed'; // Text decoration style prop
  tdc?: string; // Text decoration color prop
  tsc?: string; // Text shadow color prop
  tso?: { width: number; height: number }; // Text shadow offset prop
  tsr?: number; // Text shadow radius prop
  tt?: 'none' | 'uppercase' | 'lowercase' | 'capitalize'; // Text transform prop
  // Android-specific props
  ifp?: boolean; // Include font padding prop
  va?: 'auto' | 'top' | 'bottom' | 'middle'; // Vertical align prop
  // iOS-specific props
  wd?: 'auto' | 'ltr' | 'rtl'; // Writing direction prop
  // Universal props
  us?: 'auto' | 'text' | 'none' | 'contain' | 'all'; // User select prop
}
