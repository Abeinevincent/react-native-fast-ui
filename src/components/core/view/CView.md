# CView Component Documentation

The `CView` component is a custom view component that extends the functionality of the React Native `View` component.

## Props

### Style-related Props

- `w`: (Required) Width of the view. Accepts a DimensionValue.
- `h`: (Required) Height of the view. Accepts a DimensionValue.
- `bgc`: (Required) Background color of the view. Accepts a string representing a color code.
- `bfv`: (Optional) Backface visibility. Accepts one of: 'visible', 'hidden'.
- `bbc`: (Optional) Border bottom color. Accepts a string representing a color code.
- `bber`: (Optional) Border bottom end radius. Accepts a number.
- `bblr`: (Optional) Border bottom left radius. Accepts a number.
- `bbrr`: (Optional) Border bottom right radius. Accepts a number.
- `bbsr`: (Optional) Border bottom start radius. Accepts a number.
- `bser`: (Optional) Border start end radius. Accepts a number.
- `bssr`: (Optional) Border start start radius. Accepts a number.
- `beer`: (Optional) Border end end radius. Accepts a number.
- `besr`: (Optional) Border end start radius. Accepts a number.
- `bbw`: (Optional) Border bottom width. Accepts a number.
- `bc`: (Optional) Border color. Accepts a string representing a color code.
- `bci`: (Optional) Border curve iOS. Accepts one of: 'circular', 'continuous'.
- `bec`: (Optional) Border end color. Accepts a string representing a color code.
- `blc`: (Optional) Border left color. Accepts a string representing a color code.
- `blw`: (Optional) Border left width. Accepts a number.
- `br`: (Optional) Border radius. Accepts a number.
- `brc`: (Optional) Border right color. Accepts a string representing a color code.
- `brw`: (Optional) Border right width. Accepts a number.
- `bsc`: (Optional) Border start color. Accepts a string representing a color code.
- `bs`: (Optional) Border style. Accepts one of: 'solid', 'dotted', 'dashed'.
- `btc`: (Optional) Border top color. Accepts a string representing a color code.
- `bter`: (Optional) Border top end radius. Accepts a number.
- `btlr`: (Optional) Border top left radius. Accepts a number.
- `btrr`: (Optional) Border top right radius. Accepts a number.
- `btsr`: (Optional) Border top start radius. Accepts a number.
- `btw`: (Optional) Border top width. Accepts a number.
- `bw`: (Optional) Border width. Accepts a number.
- `el`: (Optional) Elevation Android. Sets the elevation of the view. Accepts a number.
- `o`: (Optional) Opacity. Accepts a number.
- `pe`: (Optional) Pointer events. Controls whether the view can be the target of touch events. Accepts one of: 'auto', 'none', 'box-none', 'box-only'.
- `m`: (Optional) Margin. Accepts a DimensionValue.
- `mt`: (Optional) Margin top. Accepts a DimensionValue.
- `mr`: (Optional) Margin right. Accepts a DimensionValue.
- `mb`: (Optional) Margin bottom. Accepts a DimensionValue.
- `ml`: (Optional) Margin left. Accepts a DimensionValue.
- `p`: (Optional) Padding. Accepts a DimensionValue.
- `pt`: (Optional) Padding top. Accepts a DimensionValue.
- `pr`: (Optional) Padding right. Accepts a DimensionValue.
- `pb`: (Optional) Padding bottom. Accepts a DimensionValue.
- `pl`: (Optional) Padding left. Accepts a DimensionValue.
- `ph`: (Optional) Padding horizontal. Accepts a DimensionValue.
- `pv`: (Optional) Padding vertical. Accepts a DimensionValue.
- `mv`: (Optional) Margin vertical. Accepts a DimensionValue.
- `mh`: (Optional) Margin horizontal. Accepts a DimensionValue.
- `ac`: (Optional) Align content. Accepts one of: 'flex-start', 'flex-end', 'center', 'stretch', 'space-between', 'space-around'.
- `ai`: (Optional) Align items. Accepts one of: 'flex-start', 'flex-end', 'center', 'stretch', 'baseline'.
- `as`: (Optional) Align self. Accepts one of: 'auto', 'flex-start', 'flex-end', 'center', 'stretch', 'baseline'.
- `ar`: (Optional) Aspect ratio. Accepts a number or string.
- `b`: (Optional) Bottom. Accepts a DimensionValue.
- `t`: (Optional) Top. Accepts a DimensionValue.
- `r`: (Optional) Right. Accepts a DimensionValue.
- `l`: (Optional) Left. Accepts a DimensionValue.
- `d`: (Optional) Display. Accepts one of: 'none', 'flex'.
- `e`: (Optional) End. Accepts a DimensionValue.
- `f`: (Optional) Flex. Accepts a number.
- `fb`: (Optional) Flex basis. Accepts a DimensionValue.
- `fd`: (Optional) Flex direction. Accepts one of: 'row', 'column'.
- `rg`: (Optional) Row gap. Accepts a number.
- `g`: (Optional) Gap. Accepts a number.
- `cg`: (Optional) Column gap. Accepts a number.
- `fg`: (Optional) Flex grow. Accepts a number.
- `fs`: (Optional) Flex shrink. Accepts a number.
- `fwp`: (Optional) Flex wrap. Accepts one of: 'wrap', 'nowrap', 'wrap-reverse'.
- `jc`: (Optional) Justify content. Accepts one of: 'flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'space-evenly'.
- `zi`: (Optional) Z index. Accepts a number.

### Additional Props

- `children`: (Optional) Content to be rendered inside the view.

## Example Usage

```jsx
import React from 'react';
import { CView } from './CView';

const MyComponent = () => {
  return (
    <>
      <CView w={100} h={50} bgc="blue" br={10} p={5} el={2} o={0.8} pe="auto">
        {/* Content/children goes here */}
      </CView>
      <CView
        w={100}
        h={50}
        bgc="blue"
        children={<SomeComponentToRenderAsChildren />}
      />
      <CView w={100} h={50} bgc="blue" />
    </>
  );
};

export default MyComponent;
```
