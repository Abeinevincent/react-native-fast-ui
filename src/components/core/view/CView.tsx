import React from 'react';
import { View, Platform, type ViewStyle } from 'react-native';
import type { ICView } from './CView.types';
import { StyleSheet } from 'react-native';
import { viewStyleMapper } from '../../utils/stylemapper';

export const CView: React.FC<ICView> = ({
  style: additionalStyles,
  children,
  bgc,
  ...rest
}) => {
  // Create default style object
  const viewStyle: ViewStyle = {
    backgroundColor: bgc ? bgc : 'transparent',
  };

  // Iterate through props and apply corresponding style props dynamically
  Object.keys(rest).forEach((prop) => {
    const key = prop as keyof ICView;
    if (viewStyleMapper()[key]) {
      // @ts-ignore
      viewStyle[viewStyleMapper()[key]] = rest[key];
    }
  });

  // Apply platform-specific styles
  if (Platform.OS === 'android') {
    viewStyle.elevation = rest.el ?? 0;
  }

  // Combine custom styles with additional styles
  const combinedStyles = StyleSheet.flatten([viewStyle, additionalStyles]);

  // Render CView component with combined styles and children, if provided
  return (
    <>
      <View testID="cview" {...rest} style={combinedStyles}>
        {children}
      </View>
    </>
  );
};
