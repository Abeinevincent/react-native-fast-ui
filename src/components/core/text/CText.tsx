import React from 'react';
import { Text, Platform, type TextStyle } from 'react-native';
import { type ICText } from './CText.types';
import { textStyleMapper } from '../../utils/stylemapper';

const CText: React.FC<ICText> = ({
  style: additionalStyles,
  children,
  fs,
  c,
  ...rest
}) => {
  // Create style object for default styles
  const textStyle: TextStyle = {
    fontSize: fs ? fs : 16,
    color: c ? c : 'black',
  };

  // Iterate through props and apply corresponding style props dynamically
  Object.keys(rest).forEach((prop) => {
    const key = prop as keyof ICText;
    if (textStyleMapper()[key]) {
      // @ts-ignore
      textStyle[textStyleMapper()[key]] = rest[key];
    }
  });

  // Apply platform-specific styles
  if (Platform.OS === 'android') {
    textStyle.includeFontPadding = rest.ifp ?? true;
    textStyle.verticalAlign = rest.va ?? 'auto';
  } else if (Platform.OS === 'ios') {
    textStyle.writingDirection = rest.wd ?? 'auto';
  }

  // Combine custom styles with additional styles
  const combinedStyles = [textStyle, additionalStyles];

  // Render CText component with combined styles and children
  return (
    <Text style={combinedStyles} {...rest} testID="text">
      {children}
    </Text>
  );
};

export { CText };
