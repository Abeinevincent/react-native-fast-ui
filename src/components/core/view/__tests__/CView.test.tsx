import React from 'react';
import '@testing-library/jest-native/extend-expect'; // Import this to extend Jest's expect functionality
import { render } from '@testing-library/react-native';
import { CText } from '../../text/CText';
import { CView } from '../CView';

describe('CView Component', () => {
  test('renders without crashing', () => {
    const { getByTestId } = render(
      <CView bgc="green" w={200} h={300} testID="cview" />
    );
    const cview = getByTestId('cview');
    expect(cview).toBeDefined();
  });

  test('renders children correctly', () => {
    const ChildComponent = () => <CText c={'black'}>Child Component</CText>;
    const { getByText, toJSON } = render(
      <CView bgc="green" w={200} h={300}>
        <ChildComponent />
      </CView>
    );
    const childComponent = getByText('Child Component');
    expect(childComponent).toBeDefined();
    expect(toJSON()).toMatchSnapshot();
  });

  test('applies provided styles', () => {
    const { getByTestId } = render(
      <CView
        testID="cview"
        w={100}
        h={100}
        bgc="red"
        bfv="visible"
        bbc="blue"
        bber={5}
      />
    );
    const cview = getByTestId('cview');
    expect(cview).toHaveStyle({
      width: 100,
      height: 100,
      backgroundColor: 'red',
      backfaceVisibility: 'visible',
      borderBottomColor: 'blue',
      borderBottomEndRadius: 5,
    });
  });

  // Add more test cases as needed
});
