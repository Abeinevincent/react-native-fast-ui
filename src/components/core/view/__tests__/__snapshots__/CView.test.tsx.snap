// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CView Component renders children correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "green",
      "borderRadius": 10,
      "height": 300,
      "justifyContent": "center",
      "paddingVertical": 93.38000000000001,
      "width": 200,
    }
  }
  testID="cview"
>
  <Text
    style={
      [
        {
          "color": "black",
          "fontSize": 16,
          "writingDirection": "auto",
        },
        undefined,
      ]
    }
    testID="text"
  >
    Child Component
  </Text>
</View>
`;
